AME jars shrink singularity

jproads opened the pull request:
## About the PR
AME jars can now be pried open with a crowbar and thrown into the singularity to shrink it. Once opened, the jar leaks fuel at a rate of around 60 fuel per second, and singularity-shrinking-effectiveness scales with amount of fuel left in the jar. Max effectiveness is 100 energy with 1000 fuel. For comparison, singulo lv2 is at 200 energy, 3 at 300, 4 at 600, 5 at 1000, 6 at 1500. With, say, two AME crates totalling 6 jars, an engineer can hope to neutralize a lv4 singulo at best with six well-timed throws. At lv5, 6 jars can help but it's a gamble.

This is meant to give prepared engineers a chance to neutralize what is usually a round-ending event that is easily triggered either intentionally through sabotage or unintentionally through engineering fuckups. Keyword is Prepared, because they'll have to ask for surplus AME jars from cargo and act before the singuloose gets too big, or before it eats the unopened AME jars. Creates scenarios in which engineering can choose to throw all their fuel into the singulo in the hopes of dissipating it, or to gamble on the singulo flying away and keeping the station powered. It's also risky, because you need to get within throwing range of the singulo.

Other changes include a more detailed examination message and some visual cues to indicate the jar's state. Also, jars cannot be resealed once opened.

**Screenshots**
![Screen Shot 2022-08-02 at 6 54 35 PM](https://user-images.githubusercontent.com/93178783/182358211-244c4ae1-b1b5-4a67-a8c7-955f698f7989.png)
At three stages: leftmost is closed, middle is opened and leaking fuel (with blue light), rightmost is opened and empty.
![Screen Shot 2022-08-04 at 5 08 27 PM](https://user-images.githubusercontent.com/93178783/182809280-49953836-fbed-44d3-8088-2324dc3d8736.png)


Integration:
![Screen Shot 2022-08-04 at 5 07 36 PM](https://user-images.githubusercontent.com/93178783/182809210-1803c7e0-bd24-4a3c-99aa-b1bf19679d56.png)
![Uploading Screen Shot 2022-08-04 at 5.08.43 PM.png…]()

**Changelog**


:cl:
- add: AME fuel jars can now be pried open and thrown into the singularity to shrink it. Current fuel is now also shown when examining.

mirrorcult commented:
very much like the idea of being able to neutralize the singularity with AME fuel but not sure I understand the purpose of tying it to a 'leaking' mechanic which seems a little obtuse and doesnt serve much other purpose?

jproads commented:
The leaking mechanic serves as an intermediate step between having the AME jar and it being able to neutralize the singularity. Otherwise, there's nothing stopping engineers from just stocking two crates of AMEs beside the CFG or in the PA and calling it a day. With the fuel leaking out, the window to neutralize the singulo is shorter and more skill-based, i.e. dancing around the singulo while microing each jar, can even use a pneumatic cannon to launch it from a safer distance albeit with increased delay and thus less effectiveness.

mirrorcult commented on Content.Server/AME/Components/AMEFuelContainerComponent.cs:
```diff
@@ -3,34 +3,30 @@ namespace Content.Server.AME.Components
-        }
+        public int MaxFuelAmount { get; set; }
+
+        [ViewVariables(VVAccess.ReadWrite)]
+        public bool Sealed { get; set; }
+
+        [ViewVariables(VVAccess.ReadOnly)] public int OpenFuelConsumption = 32;
+        [ViewVariables(VVAccess.ReadOnly)] public string QualityNeeded = "Prying";
```
datafield all of these (and make the properties fields if you feel like it)

mirrorcult commented on Content.Server/AME/Components/AMEFuelContainerComponent.cs:
```diff
@@ -3,34 +3,30 @@ namespace Content.Server.AME.Components
 
         protected override void Initialize()
         {
             base.Initialize();
-            _maxFuelAmount = 1000;
-            _fuelAmount = 1000;
+            MaxFuelAmount = 1000;
+            FuelAmount = 1000;
+            Sealed = true;
         }
```
this can be removed entirely, just set the defaults in the field initializer

mirrorcult commented on Content.Server/AME/AMEFuelContainerSystem.cs:
```diff
@@ -0,0 +1,160 @@
+                ("status", string.Empty)));
+        }
+
+        private void OnAMEFuelContainerInteractUsing(EntityUid uid, AMEFuelContainerComponent container,
+            InteractUsingEvent args)
+        {
+            if (!_toolSystem.HasQuality(args.Used, container.QualityNeeded))
+                return;
+
+            args.Handled = TryOpenAMEFuelContainer(uid, args.User, container);
+        }
```
iirc toolsystem has a method for `UseTool` which you should be using instead, also lets you use a doafter if you want. `RepairableSystem` uses this

mirrorcult commented on Content.Client/AME/AMEFuelContainerVisualizerSystem.cs:
```diff
@@ -0,0 +1,25 @@
+using Robust.Client.GameObjects;
+
+namespace Content.Client.AME;
+public sealed class AMEFuelContainerVisualizerSystem : VisualizerSystem<AMEFuelContainerVisualsComponent>
+{
+    protected override void OnAppearanceChange(EntityUid uid, AMEFuelContainerVisualsComponent component,
+        ref AppearanceChangeEvent args)
+    {
+        if (args.Sprite == null)
+            return;
+
+        if (args.Component.TryGetData(AMEFuelContainerVisuals.IsOpen, out bool isOpen))
+        {
+            args.Sprite.LayerSetVisible(AMEFuelContainerVisualLayers.Seal, !isOpen);
+        }
+    }
+
+    public enum AMEFuelContainerVisualLayers
+    {
+        Seal
+    }
```
this is super easy to just convert into a generic visualizer and remove all of the visualsystem/component boilerplate (all that needs to stay is the appearance data setting)

in the yaml you can do something like this:

```yaml

... # in sprite comp
    layers:
    - state: jar
    - state: seal
      map: ["seal"] # replace with just string layer key
      visible: true
... # further down

- type: Appearance
- type: GenericVisualizer
  visuals:
    enum.AMEFuelContainerVisuals.IsOpen:
      seal:
        True: { visible: false }
        False: { visible: true }
```

jproads replied:
note 4 myself: https://docs.spacestation14.io/en/engine/porting-visualizers

mirrorcult commented on Content.Server/AME/AMEFuelContainerSystem.cs:
```diff
@@ -0,0 +1,160 @@
+            if(user != null)
+                _popupSystem.PopupEntity(Loc.GetString("ame-fuel-container-component-on-pry"), uid, Filter.Entities(user.Value));
+
+            container.Dirty();
```
this isnt necessary as its serverside

mirrorcult commented on Content.Server/AME/AMEFuelContainerSystem.cs:
```diff
@@ -0,0 +1,160 @@
+        private void OnAMEFuelContainerInteractUsing(EntityUid uid, AMEFuelContainerComponent container,
+            InteractUsingEvent args)
+        {
+            if (!_toolSystem.HasQuality(args.Used, container.QualityNeeded))
+                return;
```
check that `args.Handled` isnt already true first

jproads commented:
@mirrorcult I've resolved your comments, changed some descriptions for integration (see screenies on main comment), and added a heating functionality leveraging the Artifact system. Let me know if you'd like anything else changed.

mirrorcult reviewed (changes requested):
deeeefinitely do not directly use `ArtifactComponent` and `TemperatureArtifact`, just do whatever the code for that system does.

you have an empty file called `Server` in the topmost dir

deltanedas commented:
the fuel indicator should be salvaged, might do it later
